### DISCUSSION TITLE
Mypy plugin crash - extract_python_type_from_typeengine tuple out of range.

### DISCUSSION THREAD

**Discussion by hterik** [sqlalchemy/sqlalchemy#9014] [Usage Questions]
### Describe the bug

Mypy crashes inside sqlalchemy plugin.

### To Reproduce

```python
pip install sqlalchemy[mypy] types-SQLAlchemy

SQLAlchemy==1.4.45
sqlalchemy2-stubs==0.0.2a29
types-SQLAlchemy==1.4.53.19


mypy.ini has following in plugins section: `sqlalchemy.ext.mypy.plugin`
```

### Error

```
mypy --show-traceback 
site-packages/airflow/models/taskmap.py:45: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.991
Traceback (most recent call last):
  File "mypy/semanal.py", line 6047, in accept
  File "mypy/nodes.py", line 1143, in accept
  File "mypy/semanal.py", line 1435, in visit_class_def
  File "mypy/semanal.py", line 1511, in analyze_class
  File "mypy/semanal.py", line 1538, in analyze_class_body_common
  File "mypy/semanal.py", line 1616, in apply_class_plugin_hooks
  File "site-packages/sqlalchemy/ext/mypy/plugin.py", line 258, in _base_cls_hook
    decl_class.scan_declarative_assignments_and_apply_types(ctx.cls, ctx.api)
  File "site-packages/sqlalchemy/ext/mypy/decl_class.py", line 96, in scan_declarative_assignments_and_apply_types
    _scan_declarative_assignment_stmt(
  File "site-packages/sqlalchemy/ext/mypy/decl_class.py", line 459, in _scan_declarative_assignment_stmt
    python_type_for_type = infer.infer_type_from_right_hand_nameexpr(
  File "site-packages/sqlalchemy/ext/mypy/infer.py", line 52, in infer_type_from_right_hand_nameexpr
    python_type_for_type = _infer_type_from_decl_column(
  File "site-packages/sqlalchemy/ext/mypy/infer.py", line 404, in _infer_type_from_decl_column
    python_type_for_type = extract_python_type_from_typeengine(
  File "site-packages/sqlalchemy/ext/mypy/infer.py", line 556, in extract_python_type_from_typeengine
    return get_proper_type(type_engine.args[-1])
IndexError: tuple index out of range
``` 

Dropping into pdb. If you want me to check anything else in the debugger, please let me know and i can easily do so.
```
-> return get_proper_type(type_engine.args[-1])
(Pdb) type_engine
sqlalchemy.sql.type_api.TypeEngine
(Pdb) type_engine.args
()
(Pdb) list
551  	    )
552  	    type_engine = map_instance_to_supertype(
553  	        Instance(node, []),
554  	        type_engine_sym.node,
555  	    )
556  ->	    return get_proper_type(type_engine.args[-1])
[EOF]
(Pdb) node
<TypeInfo sqlalchemy.sql.sqltypes.String>
(Pdb) node.fullname
'sqlalchemy.sql.sqltypes.String'
(Pdb) type_args
[<mypy.nodes.NameExpr object at 0x7f08badc10c0>, <mypy.nodes.NameExpr object at 0x7f08badc1140>]
(Pdb) type_engine_sym
<mypy.nodes.SymbolTableNode object at 0x7f08b34762c0>
(Pdb) type_engine
sqlalchemy.sql.type_api.TypeEngine
(Pdb) type_engine.args
```

### Versions

- OS: Ubuntu 22.04
- Python: 3.10.8
- SQLAlchemy: 1.4.45
- mypy 0.991 (compiled: yes)

### Additional context

After reading https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html
I'm a little confused if one needs both sqlalchemy[mypy] and types-SQLAlchemy. But without types-SQLAlchemy installed i get many other errors, such as `Skipping analyzing "sqlalchemy.orm.attributes": module is installed, but missing library stubs or py.typed marker  [import]`

**Comment by CaselIT** [comment]
Hi,

This issue is completely missing the code that reproduces the error. Can you provide it?

**Comment by CaselIT** [comment]
As for `types-SQLAlchemy`, that's not compatible with the mypy pligin, so you should uninstall it.

That's also probably the source of the crash

**Comment by hterik** [comment]
Thanks. Is there a way to get mypy to not suggest `types-SQLAlchemy` when running with --install-types and having sqlalchemy[mypy] already installed?

Have not been able to reproduce in isolation so can't point it to smaller section of code that triggers the fail yet, still working on it. Was hoping the stack trace alone could give some indicators to begin with.

**Comment by CaselIT** [comment]
> Thanks. Is there a way to get mypy to not suggest types-SQLAlchemy when running with --install-types and having sqlalchemy[mypy] already installed?

I don't know how mypy decides what to suggest

**Comment by zzzeek** [comment]
ive moved this to a discussion for now as it seems like it may be installation related and we also have no reproducer.    it's still not good that the plugin is mis-interpreting a mypy structure, but from our POV the mypy plugin will be hopefully no longer used within a year or so.  huge mistake to do it that way though it got us through the interim

**Comment by ddimmich** [comment]
For what its worth - I've had the same issue with sqlalchemy-stubs and sqlalchemy2-stubs installed, with sqlalchemy-stubs subsequently uninstalled.  It seems these two packages install files with the same names.  If you unsinstall sqlalchemy-stubs, you need to _reinstall_ sqlalchemy2-stubs, by doing a full uninstall and subsequently installing it again.
